python/ray/data/_internal/split.py

```diff
@@ -0,0 +1,438 @@
+import collections
+import logging
+from typing import Iterable, Tuple, List, Optional
+
+import ray
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data.block import (
+    Block,
+    BlockAccessor,
+    BlockExecStats,
+    BlockMetadata,
+)
+from ray.types import ObjectRef
+
+logger = logging.getLogger(__name__)
+
+
+def _split_at_index(
+    blocks_with_metadata: Iterable[Tuple[ObjectRef[Block], BlockMetadata]],
+    index: int,
+    return_right_half: bool,
+) -> Tuple[
+    List[ObjectRef[Block]],
+    List[BlockMetadata],
+    List[ObjectRef[Block]],
+    List[BlockMetadata],
+]:
+    """Split blocks at the provided index.
+
+    Args:
+        blocks_with_metadata: Block futures to split, including the associated metadata.
+        index: The (global) index at which to split the blocks.
+        return_right_half: Whether we want to return or drop the data to the right of
+            the index.
+    Returns:
+        The block split futures and their metadata for left and right of the index.
+    """
+    get_num_rows = cached_remote_fn(_get_num_rows)
+    split_block_at_index = cached_remote_fn(_split_block_at_index, num_returns=4)
+
+    count = 0
+    left_blocks = []
+    left_metadata = []
+    right_blocks = []
+    right_metadata = []
+    for b, m in blocks_with_metadata:
+        if m.num_rows is None:
+            num_rows = ray.get(get_num_rows.remote(b))
+        else:
+            num_rows = m.num_rows
+        if count >= index:
+            if not return_right_half:
+                break
+            right_blocks.append(b)
+            right_metadata.append(m)
+        elif count + num_rows <= index:
+            left_blocks.append(b)
+            left_metadata.append(m)
+        else:
+            b0, m0, b1, m1 = split_block_at_index.remote(
+                b, m, index - count, return_right_half
+            )
+            left_blocks.append(b0)
+            left_metadata.append(ray.get(m0))
+            right_blocks.append(b1)
+            right_metadata.append(ray.get(m1))
+        count += num_rows
+    return left_blocks, left_metadata, right_blocks, right_metadata
+
+
+def _split_at_indices(
```

Comment:
I'm wondering if we can simplify this. Like what if we broke it into two stages:
1. first figure out which splits land in which blocks
2. second execute the split tasks

That would also easier to unit test.

Reply:
+1. I also agree on this. It would be great if it can be broken into multiple stages/methods.